proposal/js-deprecations.d.ts

```diff
@@ -0,0 +1,302 @@
+/**
+ * # JavaScript Deprecations API: Draft 1
+ *
+ * *([Issue](https://github.com/sass/sass/issues/3520))*
+ *
+ * ## Background
+ *
+ * > This section is non-normative.
+ *
+ * We recently added support to Dart Sass that allowed users to opt in to
+ * treating deprecation warnings as errors (on a per-deprecation basis), as
+ * well as opting in early to certain future deprecations. This is currently
+ * supported on the command line and via the Dart API, but we'd like to extend
+ * this support to the JS API as well.
+ *
+ * We would also like to add support for silencing a particular deprecation's
+ * warnings, primarily to enable a gentler process for deprecating `@import`.
+ *
+ * ## Summary
+ *
+ * > This section is non-normative.
+ *
+ * This proposal adds two new interfaces to the JS API (`Deprecation` and
+ * `Version`), three new optional properties on `Options` (`fatalDeprecations`,
+ * `silenceDeprecations`, and `futureDeprecations`), a new parameter on
+ * `Logger.warn` (`options.deprecationType`) two type aliases (`DeprecationOrId`
+ * and `DeprecationStatus`) and a new object `deprecations` that contains the
+ * various `Deprecation` objects.
+ *
+ * All language-wide deprecations are specified in `deprecations`, and any new
+ * deprecations added in the future should update the specification accordingly.
+ * However, the `deprecations` object also allows individual implementations to
+ * add their own, implementation-specific deprecations.
```

Comment:
I think even implementation-specific deprecations should be added to this spec, since otherwise we'll either end up in a situation where typos don't produce errors or where some implementations produce errors for deprecations that are valid in other implementations.

Reply:
Done